include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -6,425 +6,762 @@
+	using return_t = int32_t;
+	static return_t __call(NBL_CONST_REF_ARG(EnumType) val)
 	{
 		using underlying_t = std::underlying_type_t<EnumType>;
-		return find_msb_helper<underlying_t>::findMSB(static_cast<underlying_t>(val));
+		return find_msb_helper<underlying_t>::__call(static_cast<underlying_t>(val));
 	}
 };
 
-#endif
+template<typename FloatingPoint>
+requires concepts::FloatingPointScalar<FloatingPoint>
+struct rsqrt_helper<FloatingPoint>
+{
+	static FloatingPoint __call(NBL_CONST_REF_ARG(FloatingPoint) x)
+	{
+		// TODO: https://stackoverflow.com/a/62239778
+		return 1.0f / std::sqrt(x);
+	}
+};
 
-template<typename Integer>
-struct find_lsb_helper;
+template<typename T>
+requires concepts::FloatingPointScalar<T>
+struct frac_helper<T>
+{
+	using return_t = T;
+	static inline return_t __call(const T x)
+	{
+		return x - std::floor(x);
+	}
+};
 
-template<>
-struct find_lsb_helper<int32_t>
+template<typename Integer>
+requires concepts::IntegralScalar<Integer>
+struct bitCount_helper<Integer>
 {
-	static int32_t findLSB(NBL_CONST_REF_ARG(int32_t) val)
+	using return_t = int32_t;
+	static return_t __call(NBL_CONST_REF_ARG(Integer) val)
 	{
-#ifdef __HLSL_VERSION
-		return spirv::findILsb(val);
-#else
-		return glm::findLSB(val);
-#endif
+		using UnsignedInteger = typename hlsl::unsigned_integer_of_size_t<sizeof(Integer)>;
+		return std::popcount(static_cast<UnsignedInteger>(val));
 	}
 };
 
-template<>
-struct find_lsb_helper<uint32_t>
+template<typename SquareMatrix>
+requires concepts::Matrix<SquareMatrix> && matrix_traits<SquareMatrix>::Square
+struct inverse_helper<SquareMatrix>
 {
-	static int32_t findLSB(NBL_CONST_REF_ARG(uint32_t) val)
+	static SquareMatrix __call(NBL_CONST_REF_ARG(SquareMatrix) mat)
 	{
-#ifdef __HLSL_VERSION
-		return spirv::findILsb(val);
-#else
-		return glm::findLSB(val);
-#endif
+		return reinterpret_cast<SquareMatrix&>(glm::inverse(reinterpret_cast<typename SquareMatrix::Base const&>(mat)));
 	}
 };
 
-#define DEFINE_FIND_LSB_COMMON_SPECIALIZATION(INPUT_INTEGER_TYPE, INTEGER_TYPE)\
-template<>\
-struct find_lsb_helper<INPUT_INTEGER_TYPE>\
-{\
-	static int32_t findLSB(NBL_CONST_REF_ARG(INPUT_INTEGER_TYPE) val)\
-	{\
-		return find_lsb_helper<INTEGER_TYPE>::findLSB(val);\
-	}\
-};\
+template<typename EnumT>
+requires std::is_enum_v<EnumT>
+struct bitCount_helper<EnumT>
+{
+	using return_t = int32_t;
+	using underlying_t = std::underlying_type_t<EnumT>;
+	static return_t __call(NBL_CONST_REF_ARG(EnumT) val)
+	{
+		return bitCount_helper<const underlying_t>::__call(reinterpret_cast<const underlying_t&>(val));
+	}
+};
 
-DEFINE_FIND_LSB_COMMON_SPECIALIZATION(int16_t, int32_t)
-DEFINE_FIND_LSB_COMMON_SPECIALIZATION(uint16_t, uint32_t)
-#ifndef __HLSL_VERSION
-DEFINE_FIND_LSB_COMMON_SPECIALIZATION(int8_t, int32_t)
-DEFINE_FIND_LSB_COMMON_SPECIALIZATION(uint8_t, uint32_t)
-#endif
+template<typename T, typename U>
+requires concepts::FloatingPoint<T> && (concepts::FloatingPoint<T> || concepts::Boolean<T>)
```

Comment:
typo in the constraint, you should be checking U not T on the `||` clause

also a compound requirement needs a `()` around the whole bool, I've had a compiler crap itself because of that once